### PR TITLE
refactored cognitive complexity

### PR DESCRIPTION
**Pull request by 1onemike**
## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ x] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist

- [ x] The code change is tested and works locally.
- [x ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x ] There is no commented out code in this PR.
- [x ] I have followed the [development checklist][dev-checklist]
- [ x] I have followed the [perfect PR recommendations][perfect-pr]
- [ x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

### CODE REVIEW

**Comment by siamzam** [comment]
How do you make sure that your refactor is successfully and has not introduced any bugs?
By adding test cases.
Refactoring without having test cases to back it is a bad practice.
Please add test cases.

File: homeassistant/components/rainmachine/__init__.py [hunk 1]
```diff
@@ -456,6 +417,169 @@ async def async_unrestrict_watering(
...
+        entry_updates["options"] = {
+            **entry.options,
+            CONF_DEFAULT_ZONE_RUN_TIME: data.pop(CONF_DEFAULT_ZONE_RUN_TIME),
+        }
+
+    if CONF_USE_APP_RUN_TIMES not in entry.options:
+        entry_updates["options"] = {**entry.options, CONF_USE_APP_RUN_TIMES: False}
+
+
+def update_config_entry(
```
Comment (siamzam) [inline homeassistant/components/rainmachine/__init__.py:496]: This is a one liner. You are introducing more complexity by moving it to a separate function.

File: homeassistant/components/rainmachine/__init__.py [hunk 2]
```diff
@@ -222,12 +222,7 @@ async def async_setup_entry(  # noqa: C901
     ip_address = entry.data[CONF_IP_ADDRESS]
 
     try:
-        await client.load_local(
-            ip_address,
-            entry.data[CONF_PASSWORD],
-            port=entry.data[CONF_PORT],
-            use_ssl=entry.data.get(CONF_SSL, DEFAULT_SSL),
-        )
+        await load_rainmachine_client(client, ip_address, entry)
```
Comment (siamzam) [inline homeassistant/components/rainmachine/__init__.py:225]: Not sure if this reduced cognitive complexity! Prefer it as it was before. It was more readable and understandable.

File: homeassistant/components/rainmachine/__init__.py [hunk 3]
```diff
@@ -236,54 +231,20 @@ async def async_setup_entry(  # noqa: C901
...
-            f"Unexpected device found at {ip_address}; expected {entry.unique_id}, "
-            f"found {controller.mac}"
-        )
+    check_unexpected_device_discovery(hass, entry, controller, ip_address)
 
     async def async_update(api_category: str) -> dict:
         """Update the appropriate API data based on a category."""
         data: dict = {}
 
         try:
```
Comment (siamzam) [inline homeassistant/components/rainmachine/__init__.py:246]: Same here. Prefer it as it was before. Perhaps move the entire async_update function outside this function.

File: homeassistant/components/rainmachine/__init__.py (hunk 1 above)
Comment (1onemike) [reply homeassistant/components/rainmachine/__init__.py:496]: Okay, will fix that.

**Pull request closed without merging**